calicoctl/calicoctl/commands/ipam/split.go

```diff
@@ -0,0 +1,208 @@
+
+import (
+	"context"
+	"fmt"
+	"math"
+	"math/big"
+	"net"
+	"strconv"
+	"strings"
+
+	"github.com/docopt/docopt-go"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	apiv3 "github.com/projectcalico/api/pkg/apis/projectcalico/v3"
+
+	"github.com/projectcalico/calico/calicoctl/calicoctl/commands/clientmgr"
+	"github.com/projectcalico/calico/calicoctl/calicoctl/commands/common"
+	"github.com/projectcalico/calico/calicoctl/calicoctl/commands/constants"
+	"github.com/projectcalico/calico/calicoctl/calicoctl/util"
+	cnet "github.com/projectcalico/calico/libcalico-go/lib/net"
+	"github.com/projectcalico/calico/libcalico-go/lib/options"
+)
+
+func Split(args []string) error {
+	doc := constants.DatastoreIntro + `Usage:
+  <BINARY_NAME> ipam split <CIDR> <NUMBER> [--config=<CONFIG>] [--allow-version-mismatch]
+
+Options:
+  -h --help                    Show this screen.
+  -c --config=<CONFIG>         Path to the file containing connection configuration in
+                               YAML or JSON format.
+                               [default: ` + constants.DefaultConfigPath + `]
+     --allow-version-mismatch  Allow client and cluster versions mismatch.
+
+Description:
+  The ipam split command splits an IP pool specified by the specified CIDR into
+  the specified number of smaller IP pools. Each child IP pool will be of equal
+  size. IP pools can only be split into a number of smaller pools that is a power
+  of 2.
+
+Examples:
+  # Split the IP pool specified by 172.0.0.0/8 into 2 smaller pools
+  calicoctl ipam split 172.0.0.0/8 2
+`
+	// Replace all instances of BINARY_NAME with the name of the binary.
+	name, _ := util.NameAndDescription()
+	doc = strings.ReplaceAll(doc, "<BINARY_NAME>", name)
+
+	parsedArgs, err := docopt.ParseArgs(doc, args, "")
+	if err != nil {
+		return fmt.Errorf("Invalid option: 'calicoctl %s'. Use flag '--help' to read about a specific subcommand.", strings.Join(args, " "))
+	}
+	if len(parsedArgs) == 0 {
+		return nil
+	}
+
+	err = common.CheckVersionMismatch(parsedArgs["--config"], parsedArgs["--allow-version-mismatch"])
+	if err != nil {
+		return err
+	}
+
+	cf := parsedArgs["--config"].(string)
+	// Get the backend client.
+	client, err := clientmgr.NewClient(cf)
+	if err != nil {
+		return err
+	}
+
+	// Check that the datastore is locked
+	ctx := context.Background()
+	locked, err := common.CheckLocked(ctx, client)
+	if err != nil {
+		return fmt.Errorf("Error while checking if datastore was locked: %s", err)
+	} else if !locked {
+		return fmt.Errorf("Datastore is not locked. Run the `calicoctl datastore migrate lock` command in order split the IP pools.")
+	}
+
+	// Find the specified IP pool.
+	poolList, err := client.IPPools().List(ctx, options.ListOptions{})
+	if err != nil {
+		return fmt.Errorf("Unable to list IP pools to find the pool specified by %s", parsedArgs["<CIDR>"])
+	}
+
+	var oldPool *apiv3.IPPool
+	for _, pool := range poolList.Items {
+		if pool.Spec.CIDR == parsedArgs["<CIDR>"] {
```

Comment:
I think we might want to support specification by name as well. 

Perhaps we need arguments:

```
--name: the name of the IP pool to split
--cidr: the CIDR of the IP pool to split
```

Or we just make the <CIDR> argument accept either a name or CIDR?